### ISSUE TITLE
Longer URL length

### ISSUE THREAD

**Issue by GongSiqiu** [axel-download-accelerator/axel#185]
Is it possible to support longer URL length?
When I download some files with long URL, axel just says it doesn't support URL longer than 1024 bits 
![axel](https://user-images.githubusercontent.com/40697840/50132884-3060b300-02c4-11e9-981d-37a5109904c8.png)

**Comment by shankar** [comment]
Yes, that would be possible but will take some effort to convert code using specific array size to dynamic allocation using malloc.

**Comment by MikeRich88** [comment]
RFC7230 says:

>   Various ad hoc limitations on request-line length are found in
>   practice.  It is RECOMMENDED that all HTTP senders and recipients
>   support, at a minimum, request-line lengths of 8000 octets.

Internet Explorer (an example of the lowest common denominator) is said to support URLs of 2048 or 2083 characters (depends on who you ask)

**Comment by ismaell** [comment]
On 10/Feb/2019 14:10, RIP GitHub wrote:
> RFC7230 says:
> 
> >   Various ad hoc limitations on request-line length are found in
> >   practice.  It is RECOMMENDED that all HTTP senders and recipients
> >   support, at a minimum, request-line lengths of 8000 octets.
> 
> Internet Explorer (an example of the lowest common denominator) is
> said to support URLs of 2048 or 2083 characters (depends on who you
> ask)

The length is currently hardcoded in the data structures, we need to use
pointers instead, we could be just taking the URL from the environment
or allocate it in the context of the struct (by adding something to
track the allocations); patches are welcome.

**Comment by itfs1992** [comment]
I got the same error, there's never been a fix for this :(

**Comment by pierobonfada** [comment]
> Is it possible to support longer URL length? When I download some files with long URL, axel just says it doesn't support URL longer than 1024 bits ![axel](https://user-images.githubusercontent.com/40697840/50132884-3060b300-02c4-11e9-981d-37a5109904c8.png)

Yes! For example, if you use Debian or Ubuntu, do this: 
1) Press Winkey, type programs or updates and config your APT repositories to also download program sources.
2) apt-get build-dep axel; apt-get source axel; dpkg --purge axel;
3) edit src/axel.h, search for MAX_STRING and change it to 2048 or anything bigger
4) in the main directory, compile the program: ./configure; make;
5) you could use a program like checkinstall (apt-get install checkinstall) to generate and install the a deb package: sudo checkinstall -D (in the main axel source directory)

So its done!

**Comment by pierobonfada** [comment]
Yes. For example, if you use Debian or Ubuntu, do this:

    1)Press Winkey, type programs or updates and config your APT repositories to also download program sources.
    2)apt-get build-dep axel; apt-get source axel; dpkg --purge axel;
    3)edit src/axel.h, search for MAX_STRING and change it to 2048 or anything bigger
    4)in the main directory, compile the program: ./configure; make;
    5)you could use a program like checkinstall (apt-get install checkinstall) to generate and install the a deb package: sudo checkinstall -D (in the main axel source directory)

**Comment by ismaell** [comment]
> Yes. For example, if you use Debian or Ubuntu, do this:
> 
> ```
> 1)Press Winkey, type programs or updates and config your APT repositories to also download program sources.
> 2)apt-get build-dep axel; apt-get source axel; dpkg --purge axel;
> 3)edit src/axel.h, search for MAX_STRING and change it to 2048 or anything bigger
> 4)in the main directory, compile the program: ./configure; make;
> 5)you could use a program like checkinstall (apt-get install checkinstall) to generate and install the a deb package: sudo checkinstall -D (in the main axel source directory)
> ```

Not really a fix.

**Comment by MikeRich88** [comment]
I don't think it would be horrible if 2048 bytes of memory were allocated instead of 1024, considering the system I'm on has 34,359,738,368 bytes at my disposal. Sounds easier than trying to deal with it dynamically.

I plug axel into youtube-dl and it fails sometimes, because of all the bullshit Google is putting into the URL.

At least in my experience, 2048 would be long enough to deal with any case where I've had it fail, on any site.

OP posted an image instead of text but eyeballing it, I'm guessing ~80 columns, ~16 lines which isn't even 1300 bytes.

**Comment by ismaell** [comment]
It should be about 16 KB extra per host, for 2 KB strings, sure it doesn't seem like a big deal, but the real problem isn't even there...

Try starting a download, cancel it, change the string size and restart that download with the new binary, best case it will crash, but it could corrupt the download silently as well. There's many issues chained to this little setting.

**Comment by foxe6** [comment]
still no resolution?

**Comment by ismaell** [comment]
> still no resolution?

Why are you expecting one?

**Comment by MikeRich88** [comment]
> Try starting a download, cancel it, change the string size and restart that download with the new binary, best case it will crash, but it could corrupt the download silently as well. There's many issues chained to this little setting.

This is an interesting problem for sure. Does the format of the .st file always use 1024 bytes to store the URL or something? (I am not directly familiar with the format)

If there is a version field then the solution is easy. If not, then maybe the solution is to use a. st2 file for new downloads, but always check for .st file for older ones. Perhaps some time later the support for .st file could be deprecated.

Well I checked .st file. No URL is stored. So now I am curious, how would changing the memory allocated for URL, cause corruption of an old download?

```
QuinnHex:Desktop mike$ hexdump -C bigfile.enc.st 
00000000  08 00 e7 45 8f 01 00 00  00 00 cc b7 2c 00 00 00  |...E........,...|
00000010  00 00 10 50 4d 08 00 00  00 00 d4 7e 75 08 00 00  |...PM......~u...|
00000020  00 00 20 a0 9a 10 00 00  00 00 a4 7f c0 10 00 00  |.. .............|
00000030  00 00 30 f0 e7 18 00 00  00 00 f4 10 09 19 00 00  |..0.............|
00000040  00 00 40 40 35 21 00 00  00 00 a4 66 71 21 00 00  |..@@5!.....fq!..|
00000050  00 00 50 90 82 29 00 00  00 00 d4 45 be 29 00 00  |..P..).....E.)..|
00000060  00 00 60 e0 cf 31 00 00  00 00 a4 d3 0f 32 00 00  |..`..1.......2..|
00000070  00 00 70 30 1d 3a 00 00  00 00 53 c0 58 3a 00 00  |..p0.:....S.X:..|
00000080  00 00 80 80 6a 42 00 00  00 00                    |....jB....|
0000008a
```

**Comment by sposmen** [comment]
Any resolution on this?
On these times having tokens for security purposes is highly demanding and cloud services like AWS S3, Azure Blob storage, Google GCS use those tokens to make those url's available. Those URLS are ver easily longer than 1024. So perhaps axel is amazing on its jobs, is getting some legacy rules that make us to look around for alternatives :cry:.

**Comment by ismaell** [comment]
You're welcome to fix the problem, or set up a bounty, etc.
